src/deltalake_ext.rs

```diff
@@ -0,0 +1,322 @@
+                    path_part.push_str("/");
+                }
+
+                path_part.push_str(k);
+                path_part.push_str("=");
+                path_part.push_str(partition_value);
+            }
+
+            format!("{}/{}", path_part, file_name)
+        } else {
+            file_name
+        };
+
+        Ok(data_path)
+    }
+}
+
+struct InMemValueIter<'a> {
+    buffer: &'a[Value],
+    current_index: usize,
+}
+
+impl<'a> InMemValueIter<'a> {
+    fn from_vec(v: &'a[Value]) -> Self {
+        Self {
+            buffer: v,
+            current_index: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for InMemValueIter<'a> {
+    type Item = Result<Value, ArrowError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let item = self.buffer.get(self.current_index);
+
+        self.current_index += 1;
+
+        item.map(|v| Ok(v.to_owned()))
+    }
+}
+
+pub fn record_batch_from_json_buffer(arrow_schema_ref: Arc<ArrowSchema>, json_buffer: &[Value]) -> Result<RecordBatch, DeltaWriterError> {
+    let row_count = json_buffer.len();
+    let mut value_ter = InMemValueIter::from_vec(json_buffer);
+    let decoder = Decoder::new(arrow_schema_ref.clone(), row_count, None);
+    let batch = decoder.next_batch(&mut value_ter)?;
+
+    // handle none
+    let batch = batch.unwrap();
+
+    Ok(batch)
+}
+
+fn extract_partition_values(partition_cols: &Vec<String>, record_batch: &RecordBatch) -> Result<HashMap<String, String>, DeltaWriterError> {
+    let mut partition_values = HashMap::new();
+
+    for col_name in partition_cols.iter() {
+        let arrow_schema = record_batch.schema();
+
+        let i = arrow_schema.index_of(col_name)?;
+        let col = record_batch.column(i); 
+
+        let partition_string = stringified_partition_value(col)?;
+
+        partition_values.insert(col_name.clone(), partition_string);
+    }
+
+    Ok(partition_values)
+}
+
+fn create_add(partition_values: &HashMap<String, String>, path: String, size: i64, num_records: i64) -> Result<Add, DeltaWriterError> {
+    let stats = Stats {
+        numRecords: num_records,
+        // TODO: calculate additional stats
+        // look at https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/kernels/aggregate.rs for pulling these stats
+        minValues: HashMap::new(),
+        maxValues: HashMap::new(),
+        nullCount: HashMap::new(),
+    };
+    let stats_string = serde_json::to_string(&stats).unwrap();
+
+    let modification_time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap();
+    let modification_time= modification_time.as_millis() as i64;
+
+    let add = Add {
+        path,
+        size,
+
+        partitionValues: partition_values.to_owned(),
+        partitionValues_parsed: None,
+
+        modificationTime: modification_time,
+        dataChange: true,
+
+        // TODO: calculate additional stats 
+        stats: Some(stats_string),
+        stats_parsed: None,
```

Comment:
can you add a todo note for also populating `stats_parsed` (https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoint-format)? I believe this is turned on in Databricks for better performance gain.

Reply:
Added `TODO` note in https://github.com/delta-io/kafka-delta-ingest/pull/1/commits/c90ff3418c7bbf0e56cce7a7dda990e3c616140a#diff-5d8c9545386ed0b9dcf9dab1722a6b484e1d3014d9a2470788dd7b20509ca642R356.